src/nvim/eval/fs.c

```diff
@@ -1409,8 +1436,13 @@ void f_writefile(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
       }
     });
   } else if (argvars[0].v_type != VAR_BLOB) {
-    semsg(_(e_invarg2),
-          _("writefile() first argument must be a List or a Blob"));
+    if (argvars[0].v_type == VAR_STRING && current_sctx.sc_sid != SID_LUA) {
+      semsg(_(e_invarg2),
+            _("writefile() does not support Vim calls if the first argument is a string"));
+    } else {
+      semsg(_(e_invarg2),
+            _("writefile() first argument must be a List or a Blob"));
+    }
     return;
```

Comment:
This doesn't work, as the error message is shown on both branches. Also according to @justinmk there is no need to use two different error messages. So keep the original error message and just change the `(argvars[0].v_type != VAR_BLOB)` condition to `(argvars[0].v_type != VAR_BLOB && !(argvars[0].v_type == VAR_STRING && current_sctx.sc_sid == SID_LUA))`.